### PR TITLE
xcodeTargets as enumerated list

### LINKED ISSUE #441: convert xcodeTarget to enumerated list
PR #353 matches multiple Xcode targets based on substring matching. Proper support means enumerating the list of targets instead. Given that it's now multiple targets, it also means that the name changes: xcodeTarget => xcodeTargets. With a breaking change like this, I think it would be wise to provide error message for the old syntax explaining how to fix it. Then remove that old syntax (and error) on a future release.

Before:

    xcodeTarget 'IOS-APP'

After

    xcodeTargets 'IOS-APP', 'WatchKit App', 'WatchKit Extension'

### PR DESCRIPTION
**Pull request by brunobowden**
Now defaults to all targets when none are specified
Corresponding unit tests updated

Fixes #441

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -48,6 +48,8 @@ j2objcConfig {
 
 Info on additional `j2objcConfig` settings are in
 [J2objcConfig.groovy](https://github.com/j2objc-contrib/j2objc-gradle/blob/master/src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcConfig.groovy#L30).
+The default will link the transpiled code in to all of your Xcode build targets. To specific
```
Comment (advayDev1) [inline README.md]: specific -> specify

File: README.md [hunk 2]
```diff
@@ -48,6 +48,8 @@ j2objcConfig {
...
+a subset, add a line for `xcodeTargets 'IOS-APP', 'IOS-APP-TESTS', 'WATCHKIT-APP', ...`
```
Comment (advayDev1) [inline README.md]: this sentence needs an ending period

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcConfig.groovy [hunk 3]
```diff
@@ -603,14 +603,21 @@ class J2objcConfig {
      */
     String xcodeProjectDir = null
     /**
-     * Xcode app target the generated library should be linked to.
+     * Xcode app targets the generated library should be linked to.
```
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcConfig.groovy]: really it is the opposite, we're linking the target to the library not the library to the target.

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcConfig.groovy [hunk 4]
```diff
@@ -603,14 +603,21 @@ class J2objcConfig {
...
      *
-     * This will automatically add linkage for any target that starts with the
-     * same name. This should include any Test and Watch targets. For the example of
-     * xcodeTarget == 'IOS-APP', it will add targets for 'IOS-APPTests',
-     * 'IOS-APP WatchKit App' & 'IOS-APP WatchKit Extension' (if they exist).
+     * This will automatically add linkage for any target in the specified list.
```
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcConfig.groovy]: Add line to javadoc:
`If empty (default), all targets defined within the Xcode project will be linked to the generated library1

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy [hunk 5]
```diff
@@ -62,9 +62,12 @@ class XcodeTask extends DefaultTask {
     @Input @Optional
     String getXcodeProjectDir() { return J2objcConfig.from(project).xcodeProjectDir }
     @Input @Optional
-    String getXcodeTarget() { return J2objcConfig.from(project).xcodeTarget }
+    List<String> getXcodeTargets() { return J2objcConfig.from(project).xcodeTargets }
 
     @Input
+    // Intermal name within Podfile that refers to debug and release pods, '-' is prohibited
```
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy]: Internal

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy [hunk 6]
```diff
@@ -281,64 +282,136 @@ class XcodeTask extends DefaultTask {
...
+        List<String> newPodfileLines = updatePodfile(
+                oldPodfileLines, xcodeTargets, podPath,
+                podNameMethod, podNameDebug, podNameRelease)
 
-        if (! xcodeTargets.contains(xcodeTarget)) {
-            throw new InvalidUserDataException(
-                    "Can't find iOS App Target '$xcodeTarget' in Podfile: ${podFile.absolutePath}")
+        // Write file only if it's changed
+        if (!oldPodfileLines.equals(newPodfileLines)) {
+            podFile.write(newPodfileLines.join("\n"))
```
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy]: single quote the \n

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy [hunk 7]
```diff
@@ -281,64 +282,136 @@ class XcodeTask extends DefaultTask {
...
-        for (xcodeTargetName in xcodeTargets) {
-            newPodFileLines = updatePodFileLines(
-                    newPodFileLines, xcodeTargetName,
-                    podNameDebug, ['Debug'], podPath)
-            newPodFileLines = updatePodFileLines(
-                    newPodFileLines, xcodeTargetName,
-                    podNameRelease, ['Release'], podPath)
+        // No targets set, then default to all
+        if (xcodeTargets.size() == 0) {
+            xcodeTargets = podfileTargets
```
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy]: add a logger.debug for the list of inferred targets.  your RE on line 273 is technically only capturing a subset of the syntactically valid target definition lines (consider for example variable amounts of white space).  i don't think you need to handle the true grammar, but at least a log line here will elucidate the issue for us when a user complains.

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy [hunk 8]
```diff
@@ -346,35 +419,42 @@ class XcodeTask extends DefaultTask {
...
+                    if (podMethodProcessed) {
                         // repeated podName lines, drop them as they should not be here
                     } else {
-                        // write updated line the first time pod is seen
-                        newPodFileLines += newPodPathLine
-                        podNameWritten = true
+                        if (addPodMethod) {
+                            // write updated line the first time pod is seen
+                            // this makes for stable in place ordering
+                            newPodfileLines += podMethodLine
```
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy]: sometimes you use plus equals other times you use .add (above).

do you want to reassign this variable each time?

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy [hunk 9]
```diff
@@ -281,64 +282,136 @@ class XcodeTask extends DefaultTask {
...
-            assert podConfiguration in ['Debug', 'Release']
-            return "'$podConfiguration'"
-        }.join(", ")
-        podConfigStr = "[$podConfigStr]"
-
-        String podMatchedLine = "pod '$podName'"
-        String newPodPathLine =
-                "pod '$podName', :configuration => $podConfigStr, :path => '$podPath'"
+        List<String> newPodfileLines = new ArrayList<>()
+        String podMethodLine = "    $podNameMethod"
```
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy]: should this be a constant?  i assume something in the method above needs to retain this same number of indentation spaces

**Comment by advayDev1** [comment]
feedback provided.  thanks!

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy (hunk 7 above)
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy]: The Podfile (and I presume the pod specs as well) are ruby files, so I'll be glad not to write a parser for that!

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy (hunk 8 above)
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy]: Replaced `+=` with `.add(...)` everywhere in the file: 3 occurrences

What do you mean by reassigning the variable?

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy (hunk 9 above)
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy]: Indenting is just for aesthetic reasons. You can see the result in the tests.

**Comment by brunobowden** [comment]
Addressed all your other comments. Replied to a handful for further discussion.

**Comment by brunobowden** [comment]
Can I get an LGTM for this?

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcConfig.groovy [hunk 10]
```diff
@@ -603,14 +603,24 @@ class J2objcConfig {
...
+     * This should include any Test and Watch targets if needed.
      */
-    String xcodeTarget = null
+    List<String> xcodeTargets = new ArrayList<>()
+    /**
+     * Add targets within Xcode to link to the shared library.
+     *
+     * If empty (default), all targets defined within the Xcode project will
```
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcConfig.groovy]: you need to move this comment to the field, not this method.
if the field is /equal/ to empty list it is true.
if the method is called with an empty list, it is not necessarily true as xcodeTargets appends its arguments, not sets them.

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy [hunk 11]
```diff
@@ -281,64 +283,142 @@ class XcodeTask extends DefaultTask {
...
+            String podNameMethod, String podNameDebug, String podNameRelease, Logger logger) {
+
+        List<String> podfileTargets = extractXcodeTargets(oldPodfileLines)
+        List<String> newPodfileLines = new ArrayList<String>(oldPodfileLines)
+
+        // No targets set, then default to all
+        if (xcodeTargets.size() == 0) {
+            xcodeTargets = podfileTargets
+            if (logger) {
+                logger.debug("xcodeTargets is default of all targes: ${podfileTargets.join(', ')}")
```
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy]: targets

**Comment by advayDev1** [comment]
lgtm with comments

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcConfig.groovy (hunk 10 above)
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcConfig.groovy]: Done

**Pull request merged**